### PR TITLE
feat(predicted_path_checker): check predicted trajectory to avoid collisions planning can not handle

### PR DESCRIPTION
**Pull request by brkay54**
## Description

Discussion: https://github.com/autowarefoundation/autoware/discussions/2859
Issue: https://github.com/autowarefoundation/autoware.universe/issues/2252

In current implementation of Autoware, control module is not responsible for collision checking, except
obstacle_collision_checker package (It can stop the vehicle just for emergency situations.). Planning module is making
stop planning and collision checking using the reference trajectory.

However, vehicles actual path would be apart from reference trajectory in some cases such as starting driving with high
lateral error and at sharp turns.

In this scenarios, planning may not handle the collision in some part of the path because it does collision checking just
on reference trajectory. I think this is a critical safety issue, because vehicle will be open to collision in some cases.

Because vehicle’s actual path is strictly depend on the controller, we decided to add new package (predicted_path_checker) to check obstacle collision on predicted path and stop the vehicle if the collision was not handled by the planning module.

image

This node will make the vehicle stop if the current situation is emergency (if there is an obstacle in brake distance.) or collision can not be handled by planning module.

**Flow Chart:**

FlowChart

**Before this PR merged, following PRs should be merged before:**
https://github.com/autowarefoundation/autoware.universe/pull/3437
https://github.com/autowarefoundation/autoware.universe/pull/3435
https://github.com/tier4/tier4_autoware_msgs/pull/82

**Proposed Solutions:**

Please feel free to share your ideas on [discussion](https://github.com/orgs/autowarefoundation/discussions/2859).
## Tests performed

Checked the vehicle state if there is an obstacle in front of the vehicle by using `planning_simulator`. It worked as I wanted.

## Notes for reviewers

While testing the package, you should also checkout these PRs:

https://github.com/autowarefoundation/autoware.universe/pull/3790
https://github.com/autowarefoundation/autoware_launch/pull/367
https://github.com/tier4/tier4_autoware_msgs/pull/88

To test the package, it is enough to run the launch file inside of the package.
## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by brkay54** [comment]
Because this node search obstacles similar like in `obstacle_stop_planner`. It can cause issue talked here: https://github.com/autowarefoundation/autoware.universe/issues/2213
Before the merge I implement same approach to this node in this PR:  https://github.com/autowarefoundation/autoware.universe/pull/2296

**Comment by brkay54** [comment]
cc @TakaHoribe 
Because the pause interface is actively used by `default_ad_api` package, when `predicted_path_checker` try to make vehicle stop, `default_ad_api` removes pause request continuously. To avoid this conflict, I want to propose that we can split the pause requests into two: "NOMINAL", "URGENT". 

pause_interface drawio

Pause interface would be run in following rules:

* Different pause request types can not removes each others pause requests. 
* Urgent pause request is more prior to nominal pause requests.

I tried this implementation and worked well. 

There are some changes for pause_interface:

- Changes in `pause_interface` -> https://github.com/brkay54/autoware.universe/tree/refactor-pause-interface
- Changes in `tier4_control_msgs` -> https://github.com/brkay54/tier4_autoware_msgs/tree/refactor-pause-interface-msgs

Because of changes in `pause_interface`, some changes are needed in `default_ad_api`:

- Changes in `default_ad_api` -> https://github.com/brkay54/autoware.universe/tree/refactor-default-ad-api

If we use pause_interface and if we want to avoid conflict (not a rarely conflict, default_ad_api is trying to remove pause request of predicted_path_checker all time), I needed to make some changes in pause_interface. What do you think about this approach?

**Comment by brkay54** [comment]
@TakaHoribe sorry to bother you, we need to decide how make the vehicle stop. I shared my proposal above, please check 🙏🏼 .

**Comment by mitsudome-r** [comment]
We had another call to discuss this. In conclusion:
* Instead of adding new Pause state, we will add `request_source` field in SetPause.srv
* Pause Interface node (vehicle_command_gate) will keep the record of requested source and only release pause whenever all the source requests `pause: false`

**Comment by brkay54** [comment]
Hello, sorry for the delay. I designed node with using point cloud for collision checking. However I needed to make some changes in design. Currently, I changed design for using predicted objects, this week I will update the PR for this.

**Comment by brkay54** [comment]
@TakaHoribe @mehmetdogru 

Hello, it is ready for review. While testing the package, you should also checkout these PRs:

https://github.com/autowarefoundation/autoware.universe/pull/3790
https://github.com/autowarefoundation/autoware_launch/pull/367
https://github.com/tier4/tier4_autoware_msgs/pull/88

To test the package, it is enough to run the launch file inside of the package.

**Review by TakaHoribe — APPROVED** [review]
LGTM!

**Comment by takayuki5168** [comment]
@brkay54 I added a label of [run-build-and-test-differential](https://github.com/autowarefoundation/autoware.universe/labels/run-build-and-test-differential) to run CI.

**Comment by TakaHoribe** [comment]
@brkay54 Can we merge this PR?

**Comment by brkay54** [comment]
Sorry for the late response, I am planning to work on this, Just i need to fix a small bug. After that, I am gonna merge it in this week.

**Comment by brkay54** [comment]
So sorry for the delay.

@mehmetdogru If it is okay for you, can we merge it?

File: launch/tier4_control_launch/launch/control.launch.py [hunk 1]
```diff
@@ -372,6 +403,7 @@ def add_launch_arg(name: str, default_value=None, description=None):
     add_launch_arg("obstacle_collision_checker_param_path")
     add_launch_arg("external_cmd_selector_param_path")
     add_launch_arg("aeb_param_path")
+    add_launch_arg("predicted_path_checker_param_path")
```
Comment (xmfcx) [inline launch/tier4_control_launch/launch/control.launch.py:406]: Probably should also add `enable_predicted_path_checker` as a launch arg here.

**Pull request merged**